### PR TITLE
Added refitting acceleration

### PR DESCRIPTION
**Pull request by cehongwang**
Added refit acceleration to existing refit pipeline. 

During the first time of compilation, the interpreter will cache the weight name mapping between weights in TRT engine and weights in state_dict. The compiler then will do a tentative refit to test whether fast refit is success or not. If not, the caching will be removed. Later on, during refitting, if this mapping cache is detected, the re-interpretation of the module is skipped. 

If the fast refit fails, the refitter falls back to the regular refit, which re-interprets the module and does refitting accordingly.

# Checklist:

- [ x] My code follows the style guidelines of this project (You can use the linters)
- [ x] I have performed a self-review of my own code
- [x ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: tests/py/dynamo/models/test_model_refit.py [hunk 1]
```diff
@@ -81,6 +81,260 @@ def test_mapping():
...
+        assertions.assertTrue(
+            torch.allclose(expected_output, refitted_output, 1e-2, 1e-2),
+            "Refit Result is not correct. Refit failed",
+        )
+        # Clean up model env
+
+    torch._dynamo.reset()
+
+
 @pytest.mark.unit
```
Comment (narendasan) [inline tests/py/dynamo/models/test_model_refit.py:338]: Add 2 tests

1. Try to refit with fast_refit but the weight map is missing
2. Try to refit with fast_refit but the weight map is wrong

File: py/torch_tensorrt/dynamo/_refit.py [hunk 2]
```diff
@@ -169,7 +230,12 @@ def refit_module_weights(
     if len(list(compiled_module.named_children())) == 0:
         inline_module = True
 
-    compiled_module = copy.deepcopy(compiled_module)
+    if not in_place:
+        if inline_module:
+            logger.warning(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:235]: Raise an exception here

File: py/torch_tensorrt/dynamo/_refit.py [hunk 3]
```diff
@@ -181,13 +247,14 @@ def refit_module_weights(
...
             if "engine" in name
         ]
-        encoded_settings = compiled_submodules[0][1].__getstate__()[0][
+        # [('_run_on_acc_0', inline_module)]
+        encoded_metadata = compiled_submodules[0][1].__getstate__()[0][
             SERIALIZED_METADATA_IDX
         ]
         assert (
-            encoded_settings != ""
+            encoded_metadata != ""
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:255]: Change the message to 

```
"The engine provided is either not refittable or was built with a version of Torch-TensorRT that is too old, please recompile using the latest version with make_refitable=True"```

File: py/torch_tensorrt/dynamo/_refit.py [hunk 4]
```diff
@@ -294,8 +362,37 @@ def refit_module_weights(
                     engine = get_engine_from_encoded_engine(
                         engine_info[ENGINE_IDX], runtime
                     )
+                    if fast_refit:
+                        encoded_metadata = compiled_submodule.__getstate__()[0][
+                            SERIALIZED_METADATA_IDX
+                        ]
+                        assert (
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:369]: Remove the assert, checked above

File: py/torch_tensorrt/dynamo/_refit.py [hunk 5]
```diff
@@ -109,45 +109,106 @@ def construct_refit_mapping(
...
-        exit(0)
+        raise AssertionError("Refitting failed.")
 
 
 def refit_module_weights(
     compiled_module: torch.fx.GraphModule | ExportedProgram,
     new_weight_module: ExportedProgram,
     inputs: Tuple[Any, ...],
     verify_output: bool = False,
+    fast_refit: bool = True,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:210]: `use_refit_cache` or `use_weight_map_cache`.

File: py/torch_tensorrt/dynamo/_refit.py [hunk 6]
```diff
@@ -294,8 +362,37 @@ def refit_module_weights(
...
+                        ]
+                        assert (
+                            encoded_metadata != ""
+                        ), "Metadata are not saved in the engine. Please recompile the engine with make_refitable=True."
+                        weight_name_map = TorchTensorRTModule.decode_metadata(
+                            encoded_metadata
+                        )["weight_name_map"]
+                        if not weight_name_map:
+                            fast_refit = False
+                            logger.warning(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:377]:
```
"This engine does not have a weight map cache, rebuilding weight map"
```

File: py/torch_tensorrt/dynamo/_refit.py [hunk 7]
```diff
@@ -294,8 +362,37 @@ def refit_module_weights(
...
+                                "Fast refitting is not supported in this module. Use regular refitting."
+                            )
             else:
                 compiled_submodule = getattr(compiled_module, name)
+                weight_name_map = None
+                if fast_refit:
+                    try:
+                        weight_name_map = compiled_submodule.weight_name_map
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:385]:
```
assert weight_name_map is not None
```

File: py/torch_tensorrt/dynamo/_refit.py [hunk 8]
```diff
@@ -294,8 +362,37 @@ def refit_module_weights(
...
+                    except AttributeError:
+                        fast_refit = False
+                        logger.warning(
+                            "You are using a old version of Torch-TensorRT. Please re-compile the engine to avoid failures."
+                        )
+                    if not weight_name_map:
+                        fast_refit = False
+                        logger.warning(
+                            "Fast refitting is not supported in this module. Use regular refitting."
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:394]: "This engine does not have a weight map cache, rebuilding weight map"

File: py/torch_tensorrt/dynamo/_refit.py [hunk 9]
```diff
@@ -294,8 +362,37 @@ def refit_module_weights(
...
+                        assert (
+                            encoded_metadata != ""
+                        ), "Metadata are not saved in the engine. Please recompile the engine with make_refitable=True."
+                        weight_name_map = TorchTensorRTModule.decode_metadata(
+                            encoded_metadata
+                        )["weight_name_map"]
+                        if not weight_name_map:
+                            fast_refit = False
+                            logger.warning(
+                                "Fast refitting is not supported in this module. Use regular refitting."
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:375]: Use -> Using.

File: py/torch_tensorrt/dynamo/_refit.py [hunk 10]
```diff
@@ -294,8 +362,33 @@ def refit_module_weights(
...
+                            )
             else:
                 compiled_submodule = getattr(compiled_module, name)
+                weight_name_map = None
+                if use_weight_map_cache:
+                    try:
+                        weight_name_map = compiled_submodule.weight_name_map
+                    except AttributeError:
+                        logger.warning(
+                            "The module was compiled wit an old version of Torch-TensorRT. Rebuilding the weight map."
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_refit.py:385]: *with

File: tests/py/dynamo/models/test_model_refit.py [hunk 11]
```diff
@@ -81,6 +81,358 @@ def test_mapping():
...
+        assertions.assertTrue(
+            torch.allclose(expected_output, refitted_output, 1e-2, 1e-2),
+            "Refit Result is not correct. Refit failed",
+        )
+        # Clean up model env
+
+    torch._dynamo.reset()
+
+
+@pytest.mark.unit
```
Comment (narendasan) [inline tests/py/dynamo/models/test_model_refit.py:278]: unittest.skipIf(torch_tensorrt.ENABLED_FEATURES.torch_tensorrt_runtime)

File: tests/py/dynamo/models/test_model_refit.py [hunk 12]
```diff
@@ -81,6 +81,358 @@ def test_mapping():
...
+            torch.allclose(expected_output, refitted_output, 1e-2, 1e-2),
+            "Refit Result is not correct. Refit failed",
+        )
+        # Clean up model env
+
+    torch._dynamo.reset()
+
+
 @pytest.mark.unit
 def test_refit_one_engine():
```
Comment (narendasan) [inline tests/py/dynamo/models/test_model_refit.py:437]: Edit the names so its clear its using the net def refit system

**Review by narendasan — APPROVED** [review]
LGTM

File: py/torch_tensorrt/dynamo/conversion/_conversion.py [hunk 13]
```diff
@@ -126,6 +126,28 @@ def convert_module(
...
+        try:
+            _refit_single_trt_engine_with_gm(
+                new_gm=module,
+                old_engine=refit_test_engine,
+                input_list=inputs,
+                settings=settings,
+                weight_name_map=interpreter_result.weight_name_map,
+            )
+        except AssertionError:
+            logger.warning("Fast refit test failed. Removing the weight map caching.")
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/_conversion.py:150]: Where's the operation that you remove the weight map caching?

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 14]
```diff
@@ -320,6 +322,141 @@ def _construct_trt_network_def(self) -> None:
...
+        """
+        Construct the weight name mapping from engine weight name to state_dict weight name.
+        Cache the weight name for future refitting usecases.
+        Two-stage weight name tracing:
+        1. Name transformation from engine weight name to state_dict weight name
+        2. Value mapping that, for each weight in INetworkDefinition search for identical weight in state_dict
+        """
+
+        def find_weight(
+            weight_name: str, np_map: dict[str, Any], sd: dict[str, Any]
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:335]: What does `np_map` mean?
